src/github.com/getlantern/detour/detour.go

```diff
@@ -0,0 +1,353 @@
+
+	// keep track of the total bytes read in this connection
+	readBytes int64
+
+	muLocalBuffer sync.Mutex
+	// localBuffer keep track of bytes sent through direct connection
+	// in initial state so we can resend them when detour
+	localBuffer bytes.Buffer
+
+	network, addr string
+	readDeadline  time.Time
+	writeDeadline time.Time
+}
+
+const (
+	stateInitial = iota
+	stateDirect
+	stateDetour
+)
+
+var statesDesc = []string{
+	"initially",
+	"directly",
+	"detoured",
+}
+
+// SetCountry sets the ISO 3166-1 alpha-2 country code to load country specific detection rules
+func SetCountry(country string) {
+	blockDetector.Store(detectorByCountry(country))
+}
+
+// Dialer returns a function with same signature of net.Dialer.Dial().
+func Dialer(d dialFunc) dialFunc {
+	return func(network, addr string) (conn net.Conn, err error) {
+		dc := &Conn{dialDetour: d, network: network, addr: addr}
+		if !whitelisted(addr) {
+			detector := blockDetector.Load().(*Detector)
+			dc.setState(stateInitial)
+			// always try direct connection first
+			dc.conn, err = net.DialTimeout(network, addr, TimeoutToDetour)
+			if err == nil {
+				if !detector.CheckConn(dc.conn) {
+					log.Tracef("Dial %s to %s succeeded", dc.stateDesc(), addr)
+					return dc, nil
+				}
+				log.Debugf("Dial %s to %s, dns hijacked, try detour", dc.stateDesc(), addr)
+			} else if detector.CheckError(err) {
+				log.Debugf("Dial %s to %s failed, try detour: %s", dc.stateDesc(), addr, err)
+			} else {
+				log.Debugf("Dial %s to %s failed: %s", dc.stateDesc(), addr, err)
+				return dc, err
+			}
+		}
+		// if whitelisted or dial directly failed, try detour
+		dc.setState(stateDetour)
+		dc.conn, err = dc.dialDetour(network, addr)
+		if err != nil {
+			log.Errorf("Dial %s to %s failed", dc.stateDesc(), addr)
+			return nil, err
+		}
+		log.Tracef("Dial %s to %s succeeded", dc.stateDesc(), addr)
+		if !whitelisted(addr) {
+			log.Tracef("Add %s to whitelist", addr)
+			addToWl(dc.addr, false)
+		}
+		return dc, err
+	}
+}
+
+// Read() implements the function from net.Conn
+func (dc *Conn) Read(b []byte) (n int, err error) {
+	if !dc.inState(stateInitial) {
+		return dc.followUpRead(b)
+	}
+	// state will always be settled after first read, safe to clear buffer at end of it
+	defer dc.resetLocalBuffer()
+	start := time.Now()
+	if !dc.readDeadline.IsZero() && dc.readDeadline.Sub(start) < 2*TimeoutToDetour {
+		log.Tracef("no time left to test %s, read %s", dc.addr, statesDesc[stateDirect])
+		dc.setState(stateDirect)
+		return dc.countedRead(b)
+	}
+	// wait for at most TimeoutToDetour to read
+	dc.getConn().SetReadDeadline(start.Add(TimeoutToDetour))
+	n, err = dc.countedRead(b)
+	dc.getConn().SetReadDeadline(dc.readDeadline)
+
+	detector := blockDetector.Load().(*Detector)
+	if err != nil {
+		log.Debugf("Error while read from %s %s: %s", dc.addr, dc.stateDesc(), err)
+		if detector.CheckError(err) {
+			return dc.detour(b)
+		}
+		return
+	}
+	if detector.CheckContent(b) {
```

Comment:
If the content that CheckContent is looking for is split across 2 reads, this won't work.